bugwarrior/services/jitamin.py

```diff
@@ -0,0 +1,240 @@
+        super().__init__(*args, **kw)
+        self.password = self.get_password("password", self.config.username)
+        self.session = requests.session()
+
+        if not self.config.verify_ssl:
+            self.session.verify = False
+            requests.packages.urllib3.disable_warnings()
+
+        response = self.session.get(self.config.base_uri + "/check")
+        if response.headers.get("Location") is None:
+            log.info("Logging into %s", self.config.base_uri)
+            soup = BeautifulSoup(response.content, "html.parser")
+            token = soup.find("input", attrs={"name": "csrf_token"})
+            if token is not None:
+                token = token.get("value")
+
+            response = self.session.post(
+                self.config.base_uri + "/check",
+                data={
+                    "username": self.config.username,
+                    "password": self.password,
+                    "remember_me": "1",
+                    "csrf_token": token,
+                },
+            )
+
+            soup = BeautifulSoup(response.content, "html.parser")
+            error = soup.find("p", attrs={"class": "alert alert-error"})
+
+            if error is not None:
+                raise ValueError(error.text)
+
+    @staticmethod
+    def get_keyring_service(config):
+        return f"jitamin://{config.base_uri}"
+
+    def get_service_metadata(self):
+        return {"url": self.config.base_uri}
+
+    def get_owner(self, issue):
+        if issue['assignee'] is not None:
+            return issue['assignee'].replace(",","").replace(" ", "_")
+
+    def issues(self):
+        for link in self.get_task_list():
+            issue = self.get_single_task(link)
+
+            if issue is None:
+                continue
+
+            extra = {
+                "url": self.config.base_uri + link,
+                "project": issue["project"],
+                "type": "task",
+                "annotations": self.build_annotations(
+                    issue["comments"], self.config.base_uri + link
+                ),
+            }
+            yield self.get_issue_for_record(issue, extra)
+
+    def get_task_list_page(self, url):
+        response = self.session.get(self.config.base_uri + url)
+
+        soup = BeautifulSoup(response.content, "html.parser")
+
+        link_list = []
+
+        for link_tag in soup.find_all("a"):
+            link = link_tag.get("href")
+            if link is None:
+                continue
+            if not "/task/" in link:
+                continue
+            if len(link.split("/")) != 5:
+                continue
+            if link in link_list:
+                continue
+            link_list.append(link)
+        return link_list
+
+    def get_task_list(self):
+        i = 1
+        while True:
+            tasklist = self.get_task_list_page(
+                "?controller=SearchController&action=index&q="
+                + self.config.query + " status:open"
```

Comment:
We typically implement the `query` option to be as open ended as possible, which usually means it would be an alternative to the default query string rather than added to it. For example, it appears that one would not be able to query for closed issues the way this is implemented.

Reply:
I first had it as an open query, but then seen, that closed issues are reopend by bugwarrior per default (see https://github.com/ralphbean/bugwarrior/blob/develop/bugwarrior/db.py#L380), only to then be closed by taskwarrior, as they have an ending, which lead to a lot of lines like this:

```
INFO:bugwarrior.db:Updating task c9de8aa8-c329-47f7-89b3-540396fdb2cc, (bw)# - root Passwort not changable vial co; status: 'completed' -> 'pending'; end: datetime.datetime(2022, 2, 25, 12, 34, tzinfo=tzutc()) -> datetime.datetime(2022, 2, 25, 12, 34, tzinfo=<UTC>)
```
every time I do `bugwarrior pull`

So instead of filling the end field, I decided to skip all closed tasks.